containers-sig-laws/app/Main.hs

```diff
@@ -0,0 +1,102 @@
+
+genSmallPairs :: (Arbitrary l, Arbitrary r) => Gen [(l, r)]
+genSmallPairs = do
+  len <- choose (0, 2)
+  replicateM len $ (,) <$> arbitrary <*> arbitrary
+
+main :: IO ()
+main = do
+  quickCheck PR.nullEmpty
+  quickCheck $ PR.nullImpliesZeroSize (genSmallPairs @Int @String)
+  quickCheck $ PR.nonZeroSizeImpliesNotNull @Int @String
+  quickCheck PR.emptyZeroSized
+  quickCheck $ PR.sizeIsNatural @Int @String
+  quickCheck $ PR.singletonOneSized @Int @String
+  quickCheck $ PR.memberEmptyFalse @Int
+  quickCheck $ PR.memberSingletonSame @Int @String
+  quickCheck $ PR.newMemberYieldsValidValue @Int @String
+  quickCheck $ PR.lookupDefaultEmpty @Int @String
+  quickCheck $ PR.listToSingleton @Int @String
+  quickCheck $ PR.singletonFromList @Int @String
+  quickCheck $ PR.keysOfSingleton @Int @String
+  quickCheck $ PR.elemsOfSingleton @Int @String
+  quickCheck $ PR.lookupMatchMember @Int @String
+
+  -- * Laws for Map.Int implementation
+  quickCheck LI.nullEmpty
+  quickCheck $ LI.nullImpliesZeroSize (genSmallPairs @Int @String)
+  quickCheck $ LI.nonZeroSizeImpliesNotNull @Int @String
+  quickCheck LI.emptyZeroSized
+  quickCheck $ LI.sizeIsNatural @Int @String
+  quickCheck $ LI.singletonOneSized @Int @String
+  quickCheck $ LI.memberEmptyFalse @Int
+  quickCheck $ LI.memberSingletonSame @Int @String
+  quickCheck $ LI.newMemberYieldsValidValue @Int @String
+  quickCheck $ LI.lookupDefaultEmpty @Int @String
+  quickCheck $ LI.listToSingleton @Int @String
+  quickCheck $ LI.singletonFromList @Int @String
+  quickCheck $ LI.keysOfSingleton @Int @String
+  quickCheck $ LI.elemsOfSingleton @Int @String
+  quickCheck $ LI.emptyInsertSingleton @Int @String
+  quickCheck $ LI.insertDeleteEmpty @Int @String
+  quickCheck $ LI.updateNothingRemoves @Int @String
+  quickCheck $ LI.updateJustId @Int @String
+  quickCheck $ LI.alterCanInsert @Int @String
+  quickCheck $ LI.lookupMatchMember @Int @String
+
+  -- * Laws for Map.Ord implementation
+  quickCheck LO.nullEmpty
+  quickCheck $ LO.nullImpliesZeroSize (genSmallPairs @Int @String)
+  quickCheck $ LO.nonZeroSizeImpliesNotNull @Int @String
+  quickCheck LO.emptyZeroSized
+  quickCheck $ LO.sizeIsNatural @String @String
+  quickCheck $ LO.singletonOneSized @String @String
+  quickCheck $ LO.memberEmptyFalse @String
+  quickCheck $ LO.memberSingletonSame @String @String
+  quickCheck $ LO.newMemberYieldsValidValue @String @String
+  quickCheck $ LO.lookupDefaultEmpty @String @String
+  quickCheck $ LO.listToSingleton @String @String
+  quickCheck $ LO.singletonFromList @String @String
+  quickCheck $ LO.keysOfSingleton @String @String
+  quickCheck $ LO.elemsOfSingleton @String @String
+  quickCheck $ LO.emptyInsertSingleton @String @String
+  quickCheck $ LO.insertDeleteEmpty @String @String
+  quickCheck $ LO.updateNothingRemoves @String @String
+  quickCheck $ LO.updateJustId @String @String
+  quickCheck $ LO.alterCanInsert @String @String
+  quickCheck $ LO.lookupMatchMember @Int @String
```

Comment:
Backpack actually allows to not only write specific properties in the generic way, but you actually can create generic test-suite by putting all tests into the single function. So In this `main` we can just write:

```
main = do
    LI.checkLaws @K @V
    LO.checkLaws @K @V
    LH.checkLaws @K @V
```

It's much easier to work with such test-suite because we don't need to remember to update every data type if we add more test-cases, we only need to add new law into group and we have tests for every data type automatically.

Reply:
But let's do this under separate issue and merge this PR first.

Reply:
Oh great! I would prefer to do it in this PR, since it will reduce the amount of the code introduced in the first place greatly, would it be fine for you also? @ChShersh